crates/neon/src/types_impl/mod.rs

```diff
@@ -362,6 +424,54 @@ impl JsString {
         }
     }
 
+    /// Convert the JavaScript String into a UTF-16 encoded [`Vec<u16>`].
+    ///
+    /// The returned vector is guaranteed to be valid UTF-16. Therefore, any external crate that
+    /// handles UTF-16 encoded strings, can assume the content to be valid and skip eventual
+    /// validation steps.
+    ///
+    /// # Example
+    ///
+    /// A function that expects a single JavaScript string as argument and prints it out as a raw
+    /// vector of `u16`s.
+    ///
+    /// ```rust
+    /// # use neon::prelude::*;
+    /// fn print_string_as_utf16(mut cx: FunctionContext) -> JsResult<JsUndefined> {
+    ///     let s = cx.argument::<JsString>(0)?.to_utf16(&mut cx);
+    ///     println!("JavaScript string as raw UTF-16: {:?}", s);
+    ///
+    ///     Ok(cx.undefined())
+    /// }
+    /// ```
+    ///
+    /// Again a function that expects a single JavaScript string as argument, but utilizes the
+    /// [`widestring`](https://crates.io/crates/widestring) crate to handle the raw [`Vec<u16>`] as
+    /// a typical string.
+    ///
+    /// ```rust
+    /// # use neon::prelude::*;
+    /// fn print_with_widestring(mut cx: FunctionContext) -> JsResult<JsUndefined> {
+    ///     let s = cx.argument::<JsString>(0)?.to_utf16(&mut cx);
+    ///     let s = unsafe { widestring::Utf16String::from_vec_unchecked(s) };
+    ///     println!("JavaScript string as UTF-16: {}", s);
+    ///
+    ///     Ok(cx.undefined())
+    /// }
+    /// ```
+    pub fn to_utf16<'a, C: Context<'a>>(&self, cx: &mut C) -> Vec<u16> {
+        let env = cx.env().to_raw();
+
+        unsafe {
+            let capacity = sys::string::utf16_len(env, self.to_raw()) + 1;
```

Comment:
I don't remember exactly how the Node-API string functions work. Do they unconditionally write a null byte at the end?

Reply:
The docs were not 100% clear about it, only talking about it in terms of truncation of insufficient buffers. So I had a peek at the source, and they indeed always add a null byte at the end.

Getting the length only doesn't include the extra byte for the null byte at the end, so we have to add `1` to the length here, or we'd lose the last character.

Reply:
Awesome, thanks! Does this mean these functions will fail if the string includes a `NULL` or do you just end up with a null terminated string that isn't actually compatible a with a C-style string (not a problem for us)?

Reply:
Reading the code it will *not* error if the string contains a `null`. The string is written without it and then a `null` is written at the end.

https://github.com/nodejs/node/blob/09ae62b9a869ff19d634b0fd1f5a17f198cd1ae7/src/js_native_api_v8.cc#L2310